File: Frontend/src/usecase/TodoUsecase/Get.ts
import { Todo } from "@/domain/entity/todo";
import type { ITodoRepository } from "../../domain/Irepository/ITodoRepository";


export class GetUsecase {
    private readonly todoRepo: ITodoRepository

    constructor(todoRepo: ITodoRepository) {
        this.todoRepo = todoRepo
    }

    async Execute(id: string): Promise<Todo> {
        const todoDTO = await this.todoRepo.FindByID(id)

        const todo = new Todo(todoDTO.todo_id,
                              todoDTO.user_id,
                              todoDTO.title,
                              todoDTO.body,
                              todoDTO.due_date,
                              todoDTO.completed_at,
                              todoDTO.created_at,
                              todoDTO.updated_at,
        )

        return todo
    }
}

File: GoAPI/src/usecase/todo/Get.go
package todo

import (
	"GoAPI/src/domain/entity"
	"GoAPI/src/domain/repository"
	"GoAPI/src/domain/valueobject"
	"fmt"
)

type GetTodo struct {
	todoRepo repository.ITodoRepository
}

func NewGetTodo(todoRepo repository.ITodoRepository) *GetTodo {
	return &GetTodo{
		todoRepo: todoRepo,
	}
}

func (gt *GetTodo) Execute(todoID valueobject.TodoID, userID valueobject.UserID) (*entity.Todo, error) {
	todo, err := gt.todoRepo.FindByID(todoID)
	if err != nil {
		return nil, fmt.Errorf("todo検索エラー:%w", err)
	}
	if todo.UserID() != userID {
		return nil, fmt.Errorf("権限がありません")
	}

	return todo, nil
}
